lib/utils/net-utils.js

```diff
@@ -0,0 +1,25 @@
+const net = require('net')
+
+const NetUtils = {
+  isPortAvailable (port, listenAddress) {
+    return new Promise((resolve) => {
+      const server = net.createServer()
+
+      server.unref()
+      server.on('error', () => resolve(false))
+
+      server.listen(port, listenAddress, () => {
+        server.close(() => resolve(true))
+      })
+    })
+  },
+
+  getAvailablePort (port, listenAddress) {
+    return NetUtils.isPortAvailable(port, listenAddress).then((available) => {
+      if (available) return Promise.resolve(port)
+      else return NetUtils.getAvailablePort(port + 1)
```

Comment:
If `config.port` starts out of bounds for ports (> 65535), then we will loop until the test times out somehow. This will appear as a hang.  I think we should at least log on the else path. Maybe we should limit the quest to say 100 calls.

Reply:
@johnjbarton there will be `RangeError` for ports start out of bounds:
```
RangeError: "port" argument must be >= 0 and < 65536
```
So I think the approach suggested above will also work properly here:
```js
if (e.code != 'EADDRINUSE' && e.code!='EACCES'){
    return reject(e);
}
```

Reply:
ok great. (I'm unclear on the state of the code here so I'll let you do the merge when ready)